File: thriving-village/backend/src/utils/cache.ts
import { Redis } from '@upstash/redis';
import { CircuitBreaker } from './circuit-breaker';

const REDIS_TIMEOUT_MS = 1500;

/**
 * Optional — if Upstash creds aren't set (e.g. local dev without them), every
 * call below degrades to "no caching" instead of throwing, so the app still
 * works without Redis configured. `signal` is called fresh per command by the
 * Upstash client, so a hung command aborts (and frees its socket) on its own
 * instead of relying on us to time it out from the outside.
 */
const redis =
  process.env.UPSTASH_REDIS_REST_URL && process.env.UPSTASH_REDIS_REST_TOKEN
    ? new Redis({
        url: process.env.UPSTASH_REDIS_REST_URL,
        token: process.env.UPSTASH_REDIS_REST_TOKEN,
        signal: () => AbortSignal.timeout(REDIS_TIMEOUT_MS),
        // The breaker is our resilience layer — the client's own retries would
        // multiply REDIS_TIMEOUT_MS several times over before a call ever
        // surfaces as a failure, defeating the fast-fail goal.
        retry: false,
      })
    : null;

/**
 * Guards every Redis call so that a slow/unreachable Upstash can't pile up
 * in-flight requests across every job/course/contest/admin-dashboard request
 * that touches the cache. Once tripped, callers fall back to "no caching"
 * (see `cached`/`invalidateScope` below) exactly like the unconfigured case.
 */
const breaker = new CircuitBreaker({
  failureThreshold: 5,
  cooldownMs: 15_000,
  halfOpenSuccesses: 2,
  maxConcurrent: 20,
});

function warnOnce(scope: string, err: unknown) {
  const message = err instanceof Error ? err.message : String(err);
  console.warn(`[cache] Redis unavailable for scope "${scope}", bypassing cache: ${message}`);
}

const NAMESPACE = 'tv';

/**
 * Cache-aside read. `scope` groups related keys (e.g. "jobs") so a write to
 * that content type can invalidate every cached read for it at once, without
 * relying on Redis SCAN (Upstash's REST API charges per command and SCAN can
 * be slow/expensive at this layer) — each cached key registers itself in a
 * small per-scope set instead.
 */
export async function cached<T>(
  scope: string,
  key: string,
  ttlSeconds: number,
  fetcher: () => Promise<T>,
): Promise<T> {
  if (!redis) return fetcher();

  const cacheKey = `${NAMESPACE}:${scope}:${key}`;

  try {
    const hit = await breaker.exec(() => redis.get<T>(cacheKey));
    if (hit !== null && hit !== undefined) return hit;
  } catch (err) {
    warnOnce(scope, err);
    return fetcher();
  }

  const value = await fetcher();
  try {
    await breaker.exec(() => redis.set(cacheKey, value, { ex: ttlSeconds }));
    await breaker.exec(() => redis.sadd(`${NAMESPACE}:${scope}:keys`, cacheKey));
  } catch (err) {
    warnOnce(scope, err);
  }
  return value;
}

/** Call after any create/update/delete affecting `scope` so the next read isn't stale. */
export async function invalidateScope(scope: string): Promise<void> {
  if (!redis) return;
  const registryKey = `${NAMESPACE}:${scope}:keys`;
  try {
    const keys = await breaker.exec(() => redis.smembers(registryKey));
    if (keys.length > 0) await breaker.exec(() => redis.del(...keys));
    await breaker.exec(() => redis.del(registryKey));
  } catch (err) {
    warnOnce(scope, err);
    // Skip invalidation — the existing short TTLs bound how stale reads get.
  }
}

File: thriving-village/backend_django/apps/core/cache.py
"""Port of backend/src/utils/cache.ts's cache-aside pattern onto django-redis.

Same per-scope key-registry approach (a Redis Set of keys per scope) so a
write to one content type can invalidate every cached read for it without an
expensive SCAN — same reasoning Strapi's version documents, kept here even
though Django talks to Upstash over the standard protocol (not the
REST API), since the registry approach is just as cheap either way and
avoids depending on `delete_pattern`'s SCAN cost at any Redis tier.

Degrades to "no caching" (calls the fetcher directly) whenever Redis is
unavailable/unconfigured (LocMemCache in local dev) or the circuit breaker is
open — mirrors the original's "never let caching block the request" rule.
"""

import logging

from django.core.cache import cache

from apps.integrations.circuit_breaker import BulkheadRejectedError, CircuitBreaker, CircuitOpenError

logger = logging.getLogger(__name__)

NAMESPACE = "tv"

_breaker = CircuitBreaker(failure_threshold=5, cooldown_seconds=15, half_open_successes=2, max_concurrent=20)


def _raw_redis():
    """Raw redis-py client when django-redis backs the cache, else None.

    The key registry is a *set* semantically — SADD/SMEMBERS/DEL are atomic
    and O(1) per member, whereas the portable fallback (get the whole pickled
    set, union, set it back) is a read-modify-write race under concurrency
    and re-serializes the entire registry on every cache write.
    """
    try:
        from django_redis import get_redis_connection

        return get_redis_connection("default")
    except Exception:  # noqa: BLE001 — LocMemCache in local dev
        return None


def _warn_once(scope: str, err: Exception) -> None:
    logger.warning('[cache] Redis unavailable for scope "%s", bypassing cache: %s', scope, err)


def cached(scope: str, key: str, ttl_seconds: int, fetcher):
    """Cache-aside read. `fetcher` is a zero-arg callable returning the value
    to cache on a miss.
    """
    cache_key = f"{NAMESPACE}:{scope}:{key}"

    try:
        hit = _breaker.exec(lambda: cache.get(cache_key))
        if hit is not None:
            return hit
    except (CircuitOpenError, BulkheadRejectedError, Exception) as err:  # noqa: BLE001
        _warn_once(scope, err)
        return fetcher()

    value = fetcher()
    registry_key = f"{NAMESPACE}:{scope}:keys"
    try:
        _breaker.exec(lambda: cache.set(cache_key, value, ttl_seconds))
        redis = _raw_redis()
        if redis is not None:
            _breaker.exec(lambda: redis.sadd(registry_key, cache_key))
        else:
            _breaker.exec(lambda: cache.set(registry_key, set(cache.get(registry_key) or set()) | {cache_key}, None))
    except (CircuitOpenError, BulkheadRejectedError, Exception) as err:  # noqa: BLE001
        _warn_once(scope, err)
    return value


def invalidate_scope(scope: str) -> None:
    """Call after any create/update/delete affecting `scope` so the next read isn't stale."""
    registry_key = f"{NAMESPACE}:{scope}:keys"
    try:
        redis = _raw_redis()
        if redis is not None:
            keys = _breaker.exec(lambda: redis.smembers(registry_key))
            if keys:
                _breaker.exec(lambda: cache.delete_many([k.decode() if isinstance(k, bytes) else k for k in keys]))
            _breaker.exec(lambda: redis.delete(registry_key))
            return
        keys = _breaker.exec(lambda: cache.get(registry_key))
        if keys:
            _breaker.exec(lambda: cache.delete_many(list(keys)))
        _breaker.exec(lambda: cache.delete(registry_key))
    except (CircuitOpenError, BulkheadRejectedError, Exception) as err:  # noqa: BLE001
        _warn_once(scope, err)
        # Skip invalidation — the existing short TTLs bound how stale reads get.
